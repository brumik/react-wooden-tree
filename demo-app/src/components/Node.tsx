import * as React from 'react';
import { CheckboxButton, CheckboxButtonOnChange } from './CheckboxButton';
import { ExpandButton, ExpandButtonOnChange } from './ExpandButton';

/**
 * Interface for the node's state property.
 */
export interface NodeState {
    checked?: boolean;
    disabled?: boolean;
    expanded?: boolean;
    selected?: boolean;
}

export interface SelectOnChange {
    (nodeId: string, selected: boolean): void;
}

export interface OnLazyLoad {
    (nodeId: string): void;
}

/**
 * Interface for all data required from the tree root.
 */
export interface ParentData {
    // Callbacks
    checkboxOnChange: CheckboxButtonOnChange;
    expandOnChange: ExpandButtonOnChange;
    selectOnChange: SelectOnChange;
    onLazyLoad: OnLazyLoad;
    showCheckbox: boolean;
    initSelectedNode: (nodeId: string) => void;

    // Icons
    showIcon: boolean;                 // < Determines if the icons are showed in nodes.
    showImage: boolean;                // < Determines if images are preferred to the icons.
    nodeIcon: string;                  // < Default icon for nodes without it.
    checkedIcon: string;               // < The checkbox-checked icon.
    uncheckedIcon: string;             // < The checkbox-unchecked icon.
    partiallyCheckedIcon: string;      // < The checkbox-partially selected icon.
    collapseIcon: string;              // < The icon for collapsing parents.
    expandIcon: string;                // < The icon for expanding parents.
    loadingIcon: string;               // < The loading icon when loading data with ajax.
    errorIcon: string;                 // < The icon displayed when lazyLoading went wrong.
    selectedIcon: string;              // < The icon for selected nodes.

    // Styling
    changedCheckboxClass: string;      // < Extra class for the changed checkbox nodes.
    selectedClass: string;             // < Extra class for the selected nodes.

    // Other
    checkboxFirst: boolean;            // < Determines the order of the icon and the checkbox.
}

/**
 * Node properties interface.
 */
export interface NodeProps {
    nodeId?: string;
    text: string;
    nodes?: NodeProps[];
    state?: NodeState;

    checkable?: boolean;
    hideCheckbox?: boolean;

    selectable?: boolean;
    selectedIcon?: string;

    lazyLoad?: boolean;
    loading?: boolean; // Null when error occurred

    attr?: {[key: string]: string};

    // Styling
    icon?: string;
    iconColor?: string;
    iconBackground?: string;
    image?: string;
    classes?: string;

    // Private
    parentData?: ParentData;
}

/**
 * @class Node
 * @extends React.Component
 *
 * Displays a node and communicates with submodules and tree.
 */
export class Node extends React.Component<NodeProps, {}> {
    /**
     * Used for default values.
     */
    public static defaultProps: NodeProps;

    /**
     * Used for determining if checkbox has changed.
     */
    private readonly defaultCheckbox: boolean;

    /**
     * Creates the Node[] components from given nodes.
     *
     * @param {NodeProps[]} nodes The nodes to render.
     * @param {ParentData} parentData The parent data to pass.
     * @returns {JSX.Element[]} The array of JSX elements with nodes.
     */
    public static renderSublist(nodes: NodeProps[], parentData: ParentData): JSX.Element[] {
        if (nodes) {
            let elements: JSX.Element[] = [];
            for (let i = 0; i < nodes.length; i++) {
                elements.push(
                    <Node
                        key={nodes[i].nodeId}
                        parentData={parentData}
                        {...nodes[i]}
                    />
                );
            }
            return elements;
        } else { return null; }
    }

    /**
     * Renders the tree element.
     *
     * @returns {JSX.Element}
     */
    public render () {
        // Indent class
        let NodeClasses = 'indent-' + this.getItemIndentSize();

        // Checkbox
        const checkbox = !this.props.hideCheckbox && this.props.parentData.showCheckbox ? (
            <CheckboxButton
                onChange={this.handleCheckChange}
                checked={this.props.state.checked}
                checkedIcon={this.props.parentData.checkedIcon}
                partiallyCheckedIcon={this.props.parentData.partiallyCheckedIcon}
                uncheckedIcon={this.props.parentData.uncheckedIcon}
            />
        ) : null;

        // Expand button: if not displayed added padding
        let openButton: JSX.Element;
        if ( ( this.props.nodes && this.props.nodes.length > 0 ) || this.props.lazyLoad ) {
            openButton = (
                <ExpandButton
                    onChange={this.handleOpenChange}
                    expanded={this.props.state.expanded}
                    loading={this.props.loading}
                    expandIcon={this.props.parentData.expandIcon}
                    collapseIcon={this.props.parentData.collapseIcon}
                    loadingIcon={this.props.parentData.loadingIcon}
                    errorIcon={this.props.parentData.errorIcon}
                />
            );
        } else {
            openButton = null;
            NodeClasses += ' NoOpenButton';
        }

        // Icon
        let icon: JSX.Element = null;
        if ( this.props.parentData.showIcon ) {

            let iconStyle: {color?: string, backgroundColor?: string} = {};

            if ( this.props.iconColor ) {
                iconStyle.color = this.props.iconColor;
            }

            if ( this.props.iconBackground ) {
                iconStyle.backgroundColor = this.props.iconBackground;
            }

            if ( this.props.parentData.showImage && this.props.image ) {
                icon = <img className={'NodeIconImage'} src={this.props.image}/>;
            } else if ( this.props.icon ) {
                icon = <i className={'Icon ' + this.props.icon} style={iconStyle}/>;
            } else {
                icon = <i className={'Icon ' + this.props.parentData.nodeIcon} style={iconStyle}/>;
            }
        }

        // Selected
        let selectedIcon: JSX.Element = null;
        if ( this.props.state.selected ) {
            if ( this.props.selectedIcon ) {
                selectedIcon = <i className={this.props.selectedIcon}/>;
            } else {
                selectedIcon = <i className={this.props.parentData.selectedIcon}/>;
            }
        }

        // Selectable class
        if ( this.props.selectable && !this.props.state.disabled ) {
            NodeClasses += ' Selectable';
        }

        // Children
        const sublist = this.props.state.expanded ? (
            Node.renderSublist(this.props.nodes, this.props.parentData)
        ) : null;

        // Determining icon order
        let icon1, icon2: JSX.Element;
        if ( this.props.parentData.checkboxFirst ) {
            icon1 = checkbox;
            icon2 = icon;
        } else {
            icon1 = icon;
            icon2 = checkbox;
        }

        // Additional classes
        if ( this.props.classes ) {
            NodeClasses += ' ' + this.props.classes;
        }
        // Changed checkbox class
        if ( this.props.state.checked !== this.defaultCheckbox ) {
            NodeClasses += ' ' + this.props.parentData.changedCheckboxClass;
        }
        // Selected class
        if ( this.props.state.selected ) {
            NodeClasses += ' ' + this.props.parentData.selectedClass;
        }

        // Data Attributes list
        let DataAttr: {[k: string]: string} = {'data-id': this.props.nodeId};

        return (
            <React.Fragment>
                <li className={NodeClasses} {...DataAttr} {...this.props.attr}>
                    {openButton}
                    {icon1}
                    {selectedIcon}
                    {icon2}
                    {/* TODO Somehow remove span but prevent change if clicked on expand or check button */}
                    <span onClick={this.handleSelected}>{this.props.text}</span>
                </li>
                {sublist}
            </React.Fragment>
        );
    }

    /**
     * Constructor.
     * @param {NodeProps} props
     */
    private constructor(props: NodeProps) {
        super(props);

        if ( this.props.state.selected ) {
            this.props.parentData.initSelectedNode(this.props.nodeId);
        }

        this.defaultCheckbox = this.props.state.checked;

        this.handleSelected = this.handleSelected.bind(this);
        this.handleCheckChange = this.handleCheckChange.bind(this);
        this.handleOpenChange = this.handleOpenChange.bind(this);
    }

    /**
     * Own checkbox handler.
     * @param {boolean} checked Contains the input field value.
     */
    private handleCheckChange(checked: boolean): void {
        if ( this.props.checkable && !this.props.state.disabled ) {
            this.props.parentData.checkboxOnChange(checked, this.props.nodeId);
        }
    }

    /**
     * Handles open event.
     * If lazy load set then loads the child nodes too.
     *
     * @param {boolean} expanded True on expand false on collapse.
     */
    private handleOpenChange(expanded: boolean): void {
        if ( this.props.lazyLoad && this.props.nodes === null ) {
            this.props.parentData.onLazyLoad(this.props.nodeId);
        }

        this.props.parentData.expandOnChange(this.props.nodeId, expanded);
    }

    /**
     * Handles the selected event. If allowed to select then calls the callback
     * function with the opposite of currently selected state.
     */
    private handleSelected(): void {
        if ( this.props.selectable && !this.props.state.disabled ) {
            this.props.parentData.selectOnChange(this.props.nodeId, !this.props.state.selected);
        }
    }

    /**
     * Returns the computed padding size for the current list item for indent.
     * @returns {number} The computed padding level.
     */
    private getItemIndentSize(): number {
        return (this.props.nodeId.split('.').length - 1);
    }
}

/**
 * Node default values.
 */
Node.defaultProps = {
    nodeId: '',
    text: '',
    nodes: null,
    state: {
        checked: false,
        expanded: false,
        disabled: false,
        selected: false
    },

    checkable: true,
    hideCheckbox: false,
    selectable: true,
    selectedIcon: null,
    lazyLoad: false,
    loading: false,

    attr: null,

    // Styling
    icon: null,
    iconColor: null,
    iconBackground: null,
    image: null,
    classes: '',

    // Private
    parentData: null,
};