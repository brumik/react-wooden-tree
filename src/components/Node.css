.indent-1 {
    padding-left: 15px
}
.indent-2 {
    padding-left: 30px
}
.indent-3 {
    padding-left: 45px
}
.indent-4 {
    padding-left: 60px
}
.indent-5 {
    padding-left: 75px
}
