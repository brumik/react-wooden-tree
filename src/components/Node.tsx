import * as React from 'react';
import { NodeProps, ParentDataType } from '..';
import { CheckboxButton } from './CheckboxButton';
import { ExpandButton } from './ExpandButton';
import { ParentDataContext } from './Tree';

/**
 * @class Node
 * @extends React.Component
 *
 * Displays a node and communicates with submodules and tree.
 */
export class Node extends React.PureComponent<NodeProps, {}> {
    /**
     * Used for default values.
     */
    public static defaultProps: NodeProps;

    /**
     * Used for determining if checkbox has changed.
     */
    private readonly defaultCheckbox: boolean;

    /**
     * Creates the Node[] components from given nodes.
     *
     * @param {string[]} nodeIds The nodes to render.
     * @param {ParentDataType} parentData The parent data to pass.
     * @returns {JSX.Element[]} The array of JSX elements with nodes.
     */
    public static renderSublist(nodeIds: string[], { connectedNode, tree }: ParentDataType): JSX.Element[] {
        let elements: JSX.Element[] = [];
        if ( connectedNode ) {
            const ConnectedNode = connectedNode;
            for (let i = 0; i < nodeIds.length; i++) {
                elements.push(
                    <ConnectedNode
                        key={nodeIds[i]}
                        nodeId={nodeIds[i]}
                    />
                );
            }
        } else {
            for (let i = 0; i < nodeIds.length; i++) {
                elements.push(
                    <Node
                        key={nodeIds[i]}
                        nodeId={nodeIds[i]}
                        {...tree[nodeIds[i]]}
                    />
                );
            }
        }
        return elements;
    }

    /**
     * Renders the tree element.
     *
     * @returns {JSX.Element}
     */
    public render (): JSX.Element {
        // Indent class
        let NodeClasses = 'indent-' + this.getItemIndentSize();

        // Checkbox
        const checkbox = !this.props.hideCheckbox && this.context.showCheckbox ? (
            <CheckboxButton
                onChange={this.handleCheckChange}
                checked={this.props.state.checked}
                checkedIcon={this.context.checkedIcon}
                partiallyCheckedIcon={this.context.partiallyCheckedIcon}
                uncheckedIcon={this.context.uncheckedIcon}
            />
        ) : null;

        // Expand button: if not displayed added padding
        let openButton: JSX.Element;
        if ( ( this.props.nodes && this.props.nodes.length > 0 ) || this.props.lazyLoad ) {
            openButton = (
                <ExpandButton
                    onChange={this.handleOpenChange}
                    expanded={this.props.state.expanded}
                    loading={this.props.loading}
                    expandIcon={this.context.expandIcon}
                    collapseIcon={this.context.collapseIcon}
                    loadingIcon={this.context.loadingIcon}
                    errorIcon={this.context.errorIcon}
                />
            );
        } else {
            openButton = null;
            NodeClasses += ' no-open-button';
        }

        // Icon
        let icon: JSX.Element = null;
        if ( this.context.showIcon ) {

            let iconStyle: {color?: string, backgroundColor?: string} = {};

            if ( this.props.iconColor ) {
                iconStyle.color = this.props.iconColor;
            }

            if ( this.props.iconBackground ) {
                iconStyle.backgroundColor = this.props.iconBackground;
            }

            if ( this.context.showImage && this.props.image ) {
                icon = <img className={'icon image'} src={this.props.image}/>;
            } else if ( this.props.icon ) {
                icon = <i className={'icon ' + this.props.icon} style={iconStyle}/>;
            } else {
                icon = <i className={'icon ' + this.context.nodeIcon} style={iconStyle}/>;
            }
        }

        // Selected
        let selectedIcon: JSX.Element = null;
        if ( this.props.state.selected ) {
            if ( this.props.selectedIcon ) {
                selectedIcon = <i className={this.props.selectedIcon}/>;
            } else {
                selectedIcon = <i className={this.context.selectedIcon}/>;
            }
        }

        // Selectable class
        if ( this.props.selectable && !this.props.state.disabled ) {
            NodeClasses += ' selectable';
        }

        // Children
        const sublist = this.props.state.expanded ? (
            Node.renderSublist(this.props.nodes, this.context)
        ) : null;

        // Determining icon order
        let icon1, icon2: JSX.Element;
        if ( this.context.checkboxFirst ) {
            icon1 = checkbox;
            icon2 = icon;
        } else {
            icon1 = icon;
            icon2 = checkbox;
        }

        // Additional classes
        if ( this.props.classes ) {
            NodeClasses += ' ' + this.props.classes;
        }
        // Changed checkbox class
        if ( this.props.state.checked !== this.defaultCheckbox ) {
            NodeClasses += ' ' + this.context.changedCheckboxClass;
        }
        // Selected class
        if ( this.props.state.selected ) {
            NodeClasses += ' ' + this.context.selectedClass;
        }

        // Data Attributes list
        let DataAttr: {[k: string]: string} = {'data-id': this.props.nodeId};

        return (
            <React.Fragment>
                <li className={NodeClasses} {...DataAttr} {...this.props.attr}>
                    {openButton}
                    {icon1}
                    {selectedIcon}
                    {icon2}
                    <span onClick={this.handleSelected} dangerouslySetInnerHTML={{ __html: this.props.text }} />
                </li>
                {sublist}
            </React.Fragment>
        );
    }

    /**
     * Constructor.
     * @param {NodeProps} props
     */
     public constructor(props: NodeProps, context: ParentDataType) {
        super(props, context);

        if ( this.props.state.selected ) {
            this.context.initSelectedNode(this.props.nodeId);
        }

        this.defaultCheckbox = this.props.state.checked;
    }

    /**
     * Own checkbox handler.
     * @param {boolean} checked Contains the input field value.
     */
    private handleCheckChange = (checked: boolean): void => {
        if ( this.props.checkable && !this.props.state.disabled ) {
            this.context.checkboxOnChange(checked, this.props.nodeId);
        }
    }

    /**
     * Handles open event.
     * If lazy load set then loads the child nodes too.
     *
     * @param {boolean} expanded True on expand false on collapse.
     */
    private handleOpenChange = (expanded: boolean): void => {
        if ( this.props.lazyLoad && this.props.nodes.length < 1 ) {
            this.context.onLazyLoad(this.props.nodeId);
        } else {
            this.context.expandOnChange(this.props.nodeId, expanded);
        }
    }

    /**
     * Handles the selected event. If allowed to select then calls the callback
     * function with the opposite of currently selected state.
     *
     * If select is disabled on the node then it works like expand.
     */
    private handleSelected = (): void => {
        if ( this.props.selectable && !this.props.state.disabled ) {
            this.context.selectOnChange(this.props.nodeId, !this.props.state.selected);
        } else {
            this.handleOpenChange(!this.props.state.expanded);
        }
    }

    /**
     * Returns the computed padding size for the current list item for indent.
     * @returns {number} The computed padding level.
     */
    private getItemIndentSize = (): number => this.props.nodeId.split('.').length - 1;
}

Node.contextType = ParentDataContext;

/**
 * Node default values.
 */
Node.defaultProps = {
    nodeId: '',
    text: '',
    nodes: [],
    state: {
        checked: false,
        expanded: false,
        disabled: false,
        selected: false
    },

    checkable: true,
    hideCheckbox: false,
    selectable: true,
    selectedIcon: null,
    lazyLoad: false,
    loading: false,

    attr: null,

    // Styling
    icon: null,
    iconColor: null,
    iconBackground: null,
    image: null,
    classes: '',
};
