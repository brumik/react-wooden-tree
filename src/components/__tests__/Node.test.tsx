import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Node, ParentData } from '../Node';

let parentData: ParentData;
const initState = {
    expanded: false,
    selected: false,
    disabled: false,
    checked: false
};
const subNodes = [
    {text: 'Sub Node 0', id: '0.0'},
    {text: 'Sub Node 1', id: '0.1'}
];

beforeEach(() => {
    parentData = {
        // Callbacks
        checkboxOnChange: null,
        expandOnChange: null,
        selectOnChange: null,
        onLazyLoad: null,
        showCheckbox: true,
        initSelectedNode: function(id: string) {},

        // Icons
        showIcon: true,
        showImage: true,
        nodeIcon: 'fa fa-ban fa-fw',
        checkedIcon: 'fa fa-check-square',
        uncheckedIcon: 'fa fa-square-o',
        partiallyCheckedIcon: 'fa fa-square',
        collapseIcon: 'fa fa-angle-down',
        expandIcon: 'fa fa-angle-right',
        loadingIcon: 'fa fa-spinner fa-spin',
        errorIcon: 'fa-exclamation-triangle',
        selectedIcon: 'fa fa-check',

        // Styling
        changedCheckboxClass: 'changed-checkbox',
        selectedClass: 'selected',

        // Other
        checkboxFirst: true,
    }
});

describe('node', () => {
    it('should render basic node', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={initState}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should not render children nodes if collapsed', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={initState}
                nodes={subNodes}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should render children nodes if expanded', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={{...initState, expanded: true}}
                nodes={subNodes}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should display unique selected icon', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={{...initState, selected: true}}
                selectedIcon={'icon icon-selected'}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should not display any icon (checkbox, icon, expand icon)', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={{...parentData, showCheckbox: false, showIcon: false}}
                id={'0'}
                state={initState}
                hideCheckbox={true}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should display loading icon', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={initState}
                loading={true}
                lazyLoad={true}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should display error lazy loading icon', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={initState}
                loading={null}
                lazyLoad={true}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should add custom class', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={initState}
                classes={'custom-class'}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should indicate selected node (icon and class)', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={{...initState, selected: true}}
            />)
            .toJSON();
        expect(node).toMatchSnapshot();
    });

    it('should indicate changed checked node (icon and class)', () => {
        const node = renderer
            .create(<Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={{...initState, checked: false}}
            />);

        // Simulate props change
        node.update(
            <Node
                text={'Rendered Node'}
                parentData={parentData}
                id={'0'}
                state={{...initState, checked: true}}
            />
        );

        expect(node.toJSON()).toMatchSnapshot();
    });
});