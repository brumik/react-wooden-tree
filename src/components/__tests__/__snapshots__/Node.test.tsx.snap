// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node should add custom class 1`] = `
<li
  className="indent-0 NoOpenButton Selectable custom-class"
  id="0"
>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should display error lazy loading icon 1`] = `
<li
  className="indent-0 Selectable"
  id="0"
>
  <button
    className="ExpandButton"
    onClick={[Function]}
  >
    <i
      className="fa-exclamation-triangle"
    />
  </button>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should display loading icon 1`] = `
<li
  className="indent-0 Selectable"
  id="0"
>
  <button
    className="ExpandButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-spinner fa-spin"
    />
  </button>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should display unique selected icon 1`] = `
<li
  className="indent-0 NoOpenButton Selectable selected"
  id="0"
>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="icon icon-selected"
  />
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should indicate changed checked node (icon and class) 1`] = `
<li
  className="indent-0 NoOpenButton Selectable changed-checkbox"
  id="0"
>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-check-square"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should indicate selected node (icon and class) 1`] = `
<li
  className="indent-0 NoOpenButton Selectable selected"
  id="0"
>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-check"
  />
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should not display any icon (checkbox, icon, expand icon) 1`] = `
<li
  className="indent-0 NoOpenButton Selectable"
  id="0"
>
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should not render children nodes if collapsed 1`] = `
<li
  className="indent-0 Selectable"
  id="0"
>
  <button
    className="ExpandButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-angle-right"
    />
  </button>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should render basic node 1`] = `
<li
  className="indent-0 NoOpenButton Selectable"
  id="0"
>
  <button
    className="SelectButton"
    onClick={[Function]}
  >
    <i
      className="fa fa-square-o"
    />
  </button>
  <i
    className="fa fa-ban fa-fw"
  />
  <span
    onClick={[Function]}
  >
    Rendered Node
  </span>
</li>
`;

exports[`node should render children nodes if expanded 1`] = `
Array [
  <li
    className="indent-0 Selectable"
    id="0"
>
    <button
        className="ExpandButton"
        onClick={[Function]}
    >
        <i
            className="fa fa-angle-down"
        />
    </button>
    <button
        className="SelectButton"
        onClick={[Function]}
    >
        <i
            className="fa fa-square-o"
        />
    </button>
    <i
        className="fa fa-ban fa-fw"
    />
    <span
        onClick={[Function]}
    >
        Rendered Node
    </span>
</li>,
  <li
    className="indent-1 NoOpenButton Selectable"
    id="0.0"
>
    <button
        className="SelectButton"
        onClick={[Function]}
    >
        <i
            className="fa fa-square-o"
        />
    </button>
    <i
        className="fa fa-ban fa-fw"
    />
    <span
        onClick={[Function]}
    >
        Sub Node 0
    </span>
</li>,
  <li
    className="indent-1 NoOpenButton Selectable"
    id="0.1"
>
    <button
        className="SelectButton"
        onClick={[Function]}
    >
        <i
            className="fa fa-square-o"
        />
    </button>
    <i
        className="fa fa-ban fa-fw"
    />
    <span
        onClick={[Function]}
    >
        Sub Node 1
    </span>
</li>,
]
`;
